service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/aws/sagemakerNotebook/ControlledAwsSagemakerNotebookResource.java

```diff
@@ -0,0 +1,206 @@
+
+public class ControlledAwsSagemakerNotebookResource extends ControlledResource {
+  private final String instanceName;
+  private final String instanceType;
+
+  @JsonCreator
+  public ControlledAwsSagemakerNotebookResource(
+      @JsonProperty("wsmResourceFields") WsmResourceFields resourceFields,
+      @JsonProperty("wsmControlledResourceFields")
+          WsmControlledResourceFields controlledResourceFields,
+      @JsonProperty("instanceName") String instanceName,
+      @JsonProperty("instanceType") String instanceType) {
+    super(resourceFields, controlledResourceFields);
+    this.instanceName = instanceName;
+    this.instanceType = instanceType;
+    validate();
+  }
+
+  private ControlledAwsSagemakerNotebookResource(
+      ControlledResourceFields common, String instanceName, String instanceType) {
+    super(common);
+    this.instanceName = instanceName;
+    this.instanceType = instanceType;
+    validate();
+  }
+
+  public ControlledAwsSagemakerNotebookResource(
+      DbResource dbResource, String instanceName, String instanceType) {
+    super(dbResource);
+    this.instanceName = instanceName;
+    this.instanceType = instanceType;
+    validate();
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  @SuppressWarnings("unchecked")
+  public <T> T castByEnum(WsmResourceType expectedType) {
+    if (getResourceType() != expectedType) {
+      throw new BadRequestException(String.format("Resource is not a %s", expectedType));
+    }
+    return (T) this;
+  }
+
+  // -- getters used in serialization --
+  @JsonProperty("wsmResourceFields")
+  public WsmResourceFields getWsmResourceFields() {
+    return super.getWsmResourceFields();
+  }
+
+  @JsonProperty("wsmControlledResourceFields")
+  public WsmControlledResourceFields getWsmControlledResourceFields() {
+    return super.getWsmControlledResourceFields();
+  }
+
+  public String getInstanceName() {
+    return instanceName;
+  }
+
+  public String getInstanceType() {
+    return instanceType;
+  }
+
+  // -- getters not included in serialization --
+
+  @Override
+  @JsonIgnore
+  public WsmResourceType getResourceType() {
+    return WsmResourceType.CONTROLLED_AWS_SAGEMAKER_NOTEBOOK;
+  }
+
+  @Override
+  @JsonIgnore
+  public WsmResourceFamily getResourceFamily() {
+    return WsmResourceFamily.AWS_SAGEMAKER_NOTEBOOK;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  @JsonIgnore
+  public Optional<UniquenessCheckAttributes> getUniquenessCheckAttributes() {
+    return Optional.of(new UniquenessCheckAttributes().uniquenessScope(UniquenessScope.GLOBAL));
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void addCreateSteps(
+      CreateControlledResourceFlight flight,
+      String petSaEmail,
+      AuthenticatedUserRequest userRequest,
+      FlightBeanBag flightBeanBag) {
+    // TODO(TERRA-312) add create steps
+    throw new ApiException("addCreateSteps NotImplemented");
```

Comment:
nit (here + elsewhere): use the more specific `NotImplementedException` from TCL instead